src/hipercontracer.1
.\" ========================================================================
.\"    _   _ _ ____            ____          _____
.\"   | | | (_)  _ \ ___ _ __ / ___|___  _ _|_   _| __ __ _  ___ ___ _ __
.\"   | |_| | | |_) / _ \ '__| |   / _ \| '_ \| || '__/ _` |/ __/ _ \ '__|
.\"   |  _  | |  __/  __/ |  | |__| (_) | | | | || | | (_| | (_|  __/ |
.\"   |_| |_|_|_|   \___|_|   \____\___/|_| |_|_||_|  \__,_|\___\___|_|
.\"
.\"      ---  High-Performance Connectivity Tracer (HiPerConTracer)  ---
.\"                https://www.nntb.no/~dreibh/hipercontracer/
.\" ========================================================================
.\"
.\" High-Performance Connectivity Tracer (HiPerConTracer)
.\" Copyright (C) 2015-2026 by Thomas Dreibholz
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.\"
.\" Contact: dreibh@simula.no
.\"
.\" ###### Setup ############################################################
.Dd October 1, 2026
.Dt hipercontracer 1
.Os hipercontracer
.\" ###### Name #############################################################
.Sh NAME
.Nm hipercontracer
.Nd High\-Performance Connectivity Tracer (HiPerConTracer)
.\" ###### Synopsis #########################################################
.Sh SYNOPSIS
.Nm hipercontracer
.Op Fl P | Fl \-ping
.br
.Op Fl T | Fl \-traceroute
.\" .Op Fl J | Fl Fl jitter
.\" .br
.Op Fl L Ar level | Fl \-loglevel Ar level
.Op Fl Z Ar on|off | Fl \-logcolor Ar on|off
.Op Fl O Ar file | Fl \-logfile Ar file
.br
.Op Fl q | Fl \-quiet
.Op Fl w | Fl \-verbose
.br
.Op Fl U Ar user|uid | Fl \-user Ar user|uid
.br
.Op Fl # Ar identifier | Fl \-measurement\-id Ar identifier
.br
.Op Fl S Ar address|\%hostname[,traffic\_class[,...]] | Fl \-source Ar address|\%hostname[,traffic\_class[,...]]
.br
.Op Fl D Ar address|hostname | Fl \-destination Ar address|hostname
.br
.Op Fl \-sources\-from\-file Ar file
.br
.Op Fl \-destinations\-from\-file Ar file
.br
.Op Fl M Ar ICMP|UDP | Fl \-iomodule Ar ICMP|UDP
.br
.Op Fl I Ar number\_\%of\_\%iterations | Fl \-iterations Ar number\_\%of\_\%iterations
.br
.Op Fl \-tracerouteinterval Ar milliseconds
.br
.Op Fl \-tracerouteintervaldeviation Ar fraction
.br
.Op Fl \-tracerouteduration Ar milliseconds
.br
.Op Fl \-tracerouterounds Ar rounds
.br
.Op Fl \-tracerouteinitialmaxttl Ar value
.br
.Op Fl \-traceroutefinalmaxttl Ar value
.br
.Op Fl \-tracerouteincrementmaxttl Ar value
.br
.Op Fl \-traceroutepacketsize Ar bytes
.br
.Op Fl \-tracerouteudpsourceport Ar port
.br
.Op Fl \-tracerouteudpdestinationport Ar port
.br
.Op Fl \-pinginterval Ar milliseconds
.br
.Op Fl \-pingintervaldeviation Ar fraction
.br
.Op Fl \-pingexpiration Ar milliseconds
.br
.Op Fl \-pingburst Ar packets
.br
.Op Fl \-pingttl Ar value
.br
.Op Fl \-pingpacketsize Ar bytes
.br
.Op Fl \-pingudpsourceport Ar port
.br
.Op Fl \-pingudpdestinationport Ar port
.br
.\" .Op Fl Fl jitterinterval Ar milliseconds
.\" .br
.\" .Op Fl Fl jitterintervaldeviation Ar fraction
.\" .br
.\" .Op Fl Fl jitterexpiration Ar milliseconds
.\" .br
.\" .Op Fl Fl jitterburst Ar packets
.\" .br
.\" .Op Fl Fl jitterttl Ar value
.\" .br
.\" .Op Fl Fl jitterpacketsize Ar bytes
.\" .br
.\" .Op Fl Fl jitterrecordraw
.\" .br
.\" .Op Fl Fl jitterudpsourceport Ar port
.\" .br
.\" .Op Fl Fl jitterudpdestinationport Ar port
.\" .br
.Op Fl R Ar directory | Fl \-resultsdirectory Ar directory
.br
.Op Fl x Ar seconds | Fl \-resultstransactionlength Ar seconds
.br
.Op Fl C Ar none|\%XZ|\%BZip2|\%GZip|\%ZLIB|\%ZSTD | Fl \-resultscompression Ar none|\%XZ|\%BZip2|\%GZip|\%ZLIB|\%ZSTD
.br
.Op Fl F Ar version | Fl \-resultsformat Ar version
.br
.Op Fl z Ar depth | Fl \-resultstimestampdepth Ar depth
.br
.Op Fl \-resultsencoding Ar HPCT|\%JSON
.br
.Op Fl \-resultsoutput Ar FILE|\%CONSOLE
.br
.Op Ar destination Ar ...
.Nm hipercontracer
.Op Fl \-check
.Nm hipercontracer
.Op Fl h | Fl \-help
.Nm hipercontracer
.Op Fl v | Fl \-version
.\" ###### Description ######################################################
.Sh DESCRIPTION
.Nm hipercontracer
is a Ping/Traceroute service. It performs regular Ping and Traceroute runs
among sites and can export the results into results statements for import into
a database.
.Pp
.\" ###### Arguments ########################################################
.Sh ARGUMENTS
The following arguments may be provided:
.Bl -tag -width indent
.It Fl P | Fl \-ping
Start the Ping service. If no service is explicitly enabled, Ping is used as default.
.It Fl T | Fl \-traceroute
Start the Traceroute service.
.\" .It Fl J | Fl Fl jitter
.\" Start the Jitter service.
.It Fl L Ar level | Fl \-loglevel Ar level
Sets the minimum logging level to the provided value: 0=trace, 1=debug, 2=info, 3=warning, 4=error, 5=fatal.
.It Fl Z Ar on|off | Fl \-logcolor Ar on|off
Enables (on) or disables (off) the usage of ANSI color escape sequences to colorise the log output according to its level.
.It Fl O Ar file | Fl \-logfile Ar file
Sets a file name for appending the log output to. If a file is set, no log output is made to the console.
Otherwise, all log output will be written to the console.
.It Fl q | Fl \-quiet
Sets the minimum logging level to 3 (warning).
.It Fl w | Fl \-verbose
Sets the minimum logging level to 0 (trace).
.It Fl U Ar user|uid | Fl \-user Ar user|uid
After startup, HiPerConTracer uses UID and GID of the given user (by name or GID).
The output directory's ownership as well as the ownership of the created results
files will be set accordingly.
.It Fl # Ar identifier | Fl \-measurement\-id Ar identifier
Sets a unique measurement identifier for the measurement. All output files will have this identifier in their file name (#<ID>).
It can be used to set a setup\-wide unique identifier to identify the vantage point. If no measurement identifier is specified,
the files will contain the process ID instead (P<ID>). Range: 0\-2147483647.
Default: 0 (using process ID instead).
.It Fl S Ar address|hostname[,traffic\_class[,...]] | Fl \-source Ar address|hostname[,traffic\_class[,...]]
Adds the given source address or hostname. It is possible to use the ANY address (0.0.0.0 or ::), in order to just use the default route.
If no traffic class is given, Best Effort (00) is used. Otherwise, the list of given traffic classes (in hexadecimal) is used. Alternatively, a traffic class can be specified by PHB name (BE, EF, AF11, AF12, AF13, AF21, AF22, AF23, AF31, AF32, AF33, AF41, AF42, AF43, CS1, CS2, CS3, CS4, CS5, CS6, CS7). In this case, the corresponding traffic class with ECN bits set to 0 is used.
Note: use lower\-case hexadecimal letters to avoid ambiguities with PHB name (e.g. be for 0xbe instead of BE PHB)!
If a hostname is provided, a DNS resolution is performed and all returned addresses are added.
If no source address is provided, HiPerConTracer automatically adds 0.0.0.0 if there is an IPv4 destination, and :: if there is an IPv6 destination given.
.It Fl D Ar address|hostname | Fl \-destination Ar address|hostname
Adds the given destination address or hostname.
If a hostname is provided, a DNS resolution is performed and all returned addresses are added.
.It Fl \-sources\-from\-file Ar file
Read sources from given file. This option may be used multiple times, to read from multiple files.
.It Fl \-destinations\-from\-file Ar file
Read destinations from given file. This option may be used multiple times, to read from multiple files.
.It Fl M Ar ICMP|UDP | Fl \-iomodule Ar ICMP|UDP
Adds an I/O module: ICMP or UDP. The option may be specified multiple times with different modules.
.It Fl I Ar number\_of\_iterations | Fl \-iterations Ar number\_of\_iterations
Limit the number of measurement iterations (measurement for all source/destination
pairs) to the given number of iterations. The default 0 lets HiPerConTracer run
continuously.
.It Fl \-tracerouteinterval Ar milliseconds
Sets the Traceroute interval (time for each full round of destinations).
.It Fl \-tracerouteintervaldeviation Ar fraction
Randomly deviate the Traceroute interval by the given fraction (as 0.0 to 1.0), that is to chose out of [interval \- fraction * interval, interval + fraction * interval], in order to avoid synchronisation effects.
Default is 0.1 (10%).
.It Fl \-tracerouteduration Ar milliseconds
Sets the Traceroute duration (timeout for each destination).
.It Fl \-tracerouterounds Ar rounds
For each Traceroute measurement, perform the given number of rounds simultaneously.
Default is 1 round. Range: 1\-64.
Each TTL test of a round will have the same ICMP checksum value ("Oslo Traceroute"),
in order to deal with load balancing in the Internet.
Different rounds will have different checksums. That is, different rounds may
experience different paths in the network.
.It Fl \-tracerouteinitialmaxttl Ar value
Start with the given maximum TTL.
Default is 6.
.It Fl \-traceroutefinalmaxttl Ar value
Finish with the given maximum TTL (when destination is not reached with current TTL setting).
Default is 36.
.It Fl \-tracerouteincrementmaxttl Ar value
Increase the maximum TTL by the given value (when destination is not reached with current TTL setting).
Default is 6.
.It Fl \-traceroutepacketsize Ar bytes
Sets the Traceroute packet size, that is IP header (20 for IPv4/40 for IPv6) + ICMP header (8)/UDP header (8) + HiPerConTracer header (16) + payload, in bytes.
The actually sent packet size always covers at least the headers for IPv4/IPv6, ICMP and HiPerConTracer. Maximum packet size is 65535.
.It Fl \-tracerouteudpsourceport Ar port
Sets the Traceroute source port for the UDP module (default: 0, for automatic allocation). Note: If using a fixed UDP port for Traceroute, different UDP source ports must be used for any other services!
.It Fl \-tracerouteudpdestinationport Ar port
Sets the Traceroute destination port for the UDP module (default: 7, for Echo).
.It Fl \-pinginterval Ar milliseconds
Sets the Ping interval (time for each full round of destinations).
Default is 5000 ms.
.It Fl \-pingintervaldeviation Ar fraction
Randomly deviate the Ping interval by the given fraction (as 0.0 to 1.0), that is to chose out of [interval \- fraction * interval, interval + fraction * interval], in order to avoid synchronisation effects.
Default is 0.1 (10%).
.It Fl \-pingexpiration Ar milliseconds
Sets the Ping duration (timeout for each destination).
Default is 10000 ms.
.It Fl \-pingburst Ar packets
For each Ping measurement, send a burst of the given number of packets.
Default is 1 packet. Range: 1\-1024.
.It Fl \-pingttl Ar value
Sets the Ping TTL value.
Default is 64. Range: 1\-255.
.It Fl \-pingpacketsize Ar bytes
Sets the Ping packet size, that is IP header (20 for IPv4/40 for IPv6) + ICMP header (8)/UDP header (8) + HiPerConTracer header (16) + payload, in bytes.
The actually sent packet size always covers at least the headers for IPv4/IPv6, transport protocol and HiPerConTracer. Default is 0 (use minimum possible value). Range: 0\-65535.
.It Fl \-pingudpsourceport Ar port
Sets the Ping source port for the UDP module (default: 0, for automatic allocation). Note: If using a fixed UDP port for Ping, different UDP source ports must be used for any other services!
.It Fl \-pingudpdestinationport Ar port
Sets the Ping destination port for the UDP module (default: 7, for Echo).
.\" .It Fl Fl jitterinterval Ar milliseconds
.\" Sets the Jitter interval (time for each full round of destinations).
.\" Default is 5000 ms.
.\" .It Fl Fl jitterintervaldeviation Ar fraction
.\" Randomly deviate the Jitter interval by the given fraction (as 0.0 to 1.0), that is to chose out of [interval - fraction * interval, interval + fraction * interval], in order to avoid synchronisation effects.
.\" Default is 0.1 (10%).
.\" .It Fl Fl jitterexpiration Ar milliseconds
.\" Sets the Jitter duration (timeout for each destination).
.\" Default is 10000 ms.
.\" .It Fl Fl jitterburst Ar packets
.\" For each Jitter measurement, send a burst of the given number of packets.
.\" Default is 16 packets. Range: 2-1024.
.\" .It Fl Fl jitterttl Ar value
.\" Sets the Jitter TTL value.
.\" Default is 64. Range: 1-255.
.\" .It Fl Fl jitterpacketsize bytes
.\" Sets the Jitter packet size, that is IP header (20 for IPv4/40 for IPv6) + ICMP header (8)/UDP header (8) + HiPerConTracer header (16) + payload, in bytes.
.\" The actually sent packet size always covers at least the headers for IPv4/IPv6, transport protocol and HiPerConTracer.
.\" Default is 128 bytes. Range: 0-65535.
.\" .It Fl Fl jitterudpsourceport Ar port
.\" Sets the Jitter source port for the UDP module (default: 0, for automatic allocation). Note: If using a fixed UDP port for Jitter, different UDP source ports must be used for any other services!
.\" .It Fl Fl jitterudpdestinationport Ar port
.\" Sets the Jitter destination port for the UDP module (default: 7, for Echo).
.\" .It Fl Fl jitterrecordraw
.\" Also record raw Ping results used for jitter computation.
.It Fl R Ar directory | Fl \-resultsdirectory Ar directory
Sets the results files directory. The results will be stored there. If not specified, to results will be stored.
.It Fl x Ar seconds | Fl \-resultstransactionlength Ar seconds
After a given number of seconds, begin a new results file.
.It Fl C Ar none|XZ|BZip2|GZip|ZLIB|ZSTD | Fl \-resultscompression Ar none|XZ|BZip2|GZip|ZLIB|ZSTD
Sets the compression for the output files.
Default: XZ.
.It Fl F Ar version | Fl \-resultsformat Ar version
Sets the results file format version.
Default: 2 (current version). Range (currently): 1\-2.
Note: A future version of HiPerConTracer may increase this default setting!
.It Fl z Ar depth | Fl \-resultstimestampdepth Ar depth
Create a timestamp\-based directory hierarchy for the results, of given depth (default: 0).
0 = none, 1 = year, 2 = year/month, 3 = year/month/day, 4 = year/month/day/hour:00, 5 = year/month/day/hour:00/hour:minute.
.It Fl \-resultsencoding Ar HPCT|\%JSON
Sets the encoding for the results. Default: HPCT (HiPerConTracer format). The setting JSON generates JSON output. Note that JSON is for custom processing only; the HiPerConTracer tools expect HPTC format.
.It Fl \-resultsoutput Ar FILE|\%CONSOLE
Sets the output for the results. Default: FILE (write to files). The setting CONSOLE writes the output to the console instead.
.It Fl \-check
Print build environment information for debugging.
.It Fl h | Fl \-help
Prints command help.
.It Fl v | Fl \-version
Prints program version.
.It Ar destination Ar ...
For convenience, any positional parameters are handled as additional \-\-destination options.
See the \-\-destination option description for details.
.El
.\" ###### Output format ####################################################
.Sh OUTPUT FORMAT
The output file is a text file (possibly compressed).
The format is as follows, depending on the measurement type:
.Bl -tag -width indent
.\" ====== Format identifier ================================================
.It Format identifier in first line
.Bl -tag -width indent
.It #? HPCT format version program
.Bl -tag -width indent
.It * format: Format of the output data (Ping, Traceroute, etc.).
.It * version: Version of the output data format (decimal).
.It * program: Identifier for the program generating the output (e.g. HiPerConTracer/2.1.12).
.El
.El
.\" ====== Ping, version 2 ==================================================
.It Ping (version 2, current)
Each Ping entry consists of a #P<m> line, with m=i for ICMP and m=u for UDP (according to underlying I/O module).
.Bl -tag -width indent
.It #P<m> measurementID sourceIP destinationIP timestamp burstseq traffic\_class packetsize response\_size checksum sourcePort destinationPort status timesource delay\_app\_send delay\_queuing delay\_app\_receive rtt\_app rtt\_sw rtt\_hw
.Bl -tag -width indent
.It * measurementID: Measurement identifier.
.It * sourceIP: Source IP address.
.It * destinationIP: Destination IP address.
.It * sendTimestamp: Send Timestamp (nanoseconds since the UTC epoch, hexadecimal).
.It * burstseq: Sequence number within a burst (decimal), numbered from 0.
.It * traffic\_class: The IP Traffic Class/Type of Service value of the sent packets (hexadecimal).
.It * packet\_size: The sent packet size (decimal, in bytes) including IPv4/IPv6 header, transport header and HiPerConTracer header.
.It * response\_size: The response packet size (decimal, in bytes) including IPv4/IPv6 header, transport header and HiPerConTracer header.
.It * checksum: The checksum of the ICMP Echo Request packets (hexadecimal); 0x0000 for other protocols, 0xffff for unknown.
.It * sourcePort: Source port, 0 for ICMP (decimal).
.It * destinationPort: Destination port, 0 for ICMP (decimal).
.It * status: Status code (decimal):
.Bl -tag -width indent
.It \- 1 = ICMP response: Time Exceeded
.It \- 100 = ICMP response: Unreachable scope
.It \- 101 = ICMP response: Unreachable network
.It \- 102 = ICMP response: Unreachable host
.It \- 103 = ICMP response: Unreachable protocol
.It \- 104 = ICMP response: Unreachable port
.It \- 105 = ICMP response: Unreachable, prohibited (firewall)
.It \- 110 = ICMP response: Unreachable, unknown reason
.It \- 200 = Timeout (no response from a router)
.It \- 210 = sendto() call failed (generic error)
.It \- 211 = sendto() error: tried to send to broadcast address (EACCES)
.It \- 212 = sendto() error: network unreachable (ENETUNREACH)
.It \- 213 = sendto() error: host unreachable (EHOSTUNREACH)
.It \- 214 = sendto() error: address not available (EADDRNOTAVAIL)
.It \- 215 = sendto() error: invalid message size (EMSGSIZE)
.It \- 216 = sendto() error: not enough buffer space (ENOBUFS)
.It \- 255 = Success (destination has responded)
.El
Status values from 100 to 199 denote unreachability, as reported by ICMP error from a router. The rtt\_* fields contain the response time of the corresponding router.
Status values from 200 to 254 have the destination IP in their destinationIP field. However, there was no response from a router, so the rtt\_* fields will be the configured timeout value (on timeout) or send trial time (on send errors)!
.It * timesource: Source of the timing information (hexadecimal) as: AAQQSSHH
.Bl -tag -width indent
.It AA: Application
.It QQ: Queuing (queuing packet until sending it by driver, in software)
.It SS: Software (sending request by driver until receiving response by driver, in software; decimal, in nanoseconds; \-1 if not available)
.It HW: Hardware (sending request by NIC until receiving response by NIC, in hardware; decimal, in nanoseconds; \-1 if not available)
.El
Each byte AA, QQ, SS, HH provides the receive time source (upper nibble) and send time source (lower nibble):
.Bl -tag -width indent
.It 0x0: Not available
.It 0x1: System clock

ChangeLog
commit 28dd8a99ba1155d32c21e7c7feeb6368c826334b
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Oct 6 14:26:45 2026 +0200

    Minor CMakeLists.txt improvements.

commit d81e0c708c430bda379bdbd828064951bdd5d2a1
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Oct 6 13:52:57 2026 +0200

    Minor documentation update.

commit 66fd3332efc4fccab5714dd1a7cd9e5fd041953b
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Oct 6 11:53:06 2026 +0200

    Improved documentation.

commit 9dcae89721061f15856f2a91ae30d2b038c1e3e1
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Oct 6 10:50:17 2026 +0200

    CMakeLists.txt improvement.

commit ea43db171223a6120ccd7de80f67052a1f68a4e0
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Oct 5 22:44:40 2026 +0200

    CMakeLists.txt improvements.

commit 5ccf3cd7e446f62bd8c1ef866ebab7f62b329645
Merge: 35e47926 fc0c96dd
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Oct 5 16:28:52 2026 +0200

    Merge branch 'dreibh/json-output'

commit 35e47926c9839061cfa70c40e248f6d3ef5c4d81
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Oct 5 15:25:24 2026 +0100

    Added missing non-inotify importer part for GNU Hurd.

commit fc0c96dd29914b49aeb7d6e841bd57473981bc60
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Oct 5 14:02:47 2026 +0200

    Packaging improvement for Debian/Ubuntu/Fedora/OpenSuSE: Moved hipercontracer-examples from "Recommends" to "Suggests", to avoid dependency-heavy R installation by default when just installing "hipercontracer".

commit 0c67b7959488df8fdc6f503463f0f72921504a17
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Oct 5 11:29:54 2026 +0200

    Improved documentation.

commit 0eb15d0f21b183e6e0f855632951a31f2ae92ce9
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Oct 5 11:17:18 2026 +0200

    Minor documentation update.

commit c160585c31654172bb26dc284c78c3db062d1185
Merge: 465e2366 69e7a3c7
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Oct 5 11:12:18 2026 +0200

    Merge branch 'master' into dreibh/json-output

commit 6c2d531b99fbe48b4e7fd64bb33992af03f4f669
Merge: c6bec153 69e7a3c7
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sun Oct 4 20:08:12 2026 +0100

    Merge branch 'master' into dreibh/hurd

commit c6bec153452db5c51e93ecdf5641e8993a9d27df
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sun Oct 4 20:07:21 2026 +0100

    Minor GNU Hurd improvements and clean-ups.

commit 69e7a3c7ff316a5f45c037135386501b54155018
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sun Oct 4 21:00:58 2026 +0200

    Typo fixes.

commit 3affa18fb8e5efd16d52d85eba8fcd673082de4b
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat Oct 3 20:30:07 2026 +0100

    GNU Hurd does not support IPV6_TCLASS, yet.

commit ab2a97aa70ab47827e3b910e8223803d106be3b7
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat Oct 3 20:05:26 2026 +0100

    Added experimental support for GNU Hurd.

commit 465e236613e90e296072c92a9cfb61640d7435be
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Oct 2 17:35:23 2026 +0200

    Updated build helper script.

commit ae94c4a5efbbc77538fa1b6b9f78d912baad033c
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Oct 2 15:35:23 2026 +0200

    Type-hinting fixes in the Python examples.

commit b1389d3f9fc54c8e7646909e474394a6bccf1cef
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Oct 2 15:35:14 2026 +0200

    Updated helper scripts.

commit 1255362e24b7b08621b5c9f98dd1fd8aa9de7cfe
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Oct 2 15:20:25 2026 +0200

    Improved dependency installation hints.

commit 2dbb4766604bb15625251fe1be93646b39cd062b
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Oct 1 16:25:19 2026 +0200

    Improved script.

commit 1d17f11ec6b367f1e14a75e8041bdc416daaa5f4
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Oct 1 14:04:28 2026 +0200

    Updated HPCT-Trigger tool.

commit 83b26f7c0e24ca86ad4d7099fdcbb9972b8c9c78
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Oct 1 13:51:25 2026 +0200

    Updated shell completion.

commit 4b0700fd0aa54c59a1a25e9079992f270b1bfe06
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Oct 1 13:51:14 2026 +0200

    Updated manpage.

commit 7090f0e4f83c61d0e181c6ad35375e9e7bf6da36
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Oct 1 13:13:48 2026 +0200

    Improved mutex handling.

commit 85876b461fe12a598f4548ec79ec2d28e9da40a1
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Oct 1 13:13:36 2026 +0200

    Updated examples.

commit 7ea1421068192ff648b94072cb18e7d5bbf753ad
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Oct 1 12:35:02 2026 +0200

    Some clean-ups.

commit ee71ffa56969e54f681a5539c44d991c60ebf96b
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Oct 1 12:23:37 2026 +0200

    Changed JSON output from 3-spaces indentation to tab indentation.

commit 0387c29c81749744f7ae63a893a08449798cf4ee
Author: Alojz Gomola <alojz@simula.no>
Date:   Mon Sep 28 13:31:30 2026 +0200

    Change Request: Hipercontracer outputing to json -> targetting console/file.

commit c7fd5a68cea45b9e39438935d3f353edbf559048
Author: Alojz Gomola <alojz@simula.no>
Date:   Mon Sep 28 13:25:11 2026 +0200

    Added JSON results encoding and console output.
    
    - New options: --resultsencoding HPCT|JSON, --resultsoutput FILE|CONSOLE.
    - JSON (snake_case fields) for Ping and Traceroute.
    - Default output (HPCT files) unchanged.
    
    Co-Authored-By: Claude Opus 5.5 (1M context) <noreply@anthropic.com>

commit bf77b7a9345393cb60d709dd1a90f5033f99b615
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat Sep 12 22:44:00 2026 +0200

    OpenSUSE RPM spec fix.

commit b4a614fe26ce63b25ea73cfb9c626e29ac7bc88b
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat Sep 12 16:41:30 2026 +0100

    Updated build helper script.

commit b61b6e900bbccaccf422b53929fa921d395d31ba

src/python/HiPerConTracer.py
#!/usr/bin/env python3
# ==========================================================================
#     _   _ _ ____            ____          _____
#    | | | (_)  _ \ ___ _ __ / ___|___  _ _|_   _| __ __ _  ___ ___ _ __
#    | |_| | | |_) / _ \ '__| |   / _ \| '_ \| || '__/ _` |/ __/ _ \ '__|
#    |  _  | |  __/  __/ |  | |__| (_) | | | | || | | (_| | (_|  __/ |
#    |_| |_|_|_|   \___|_|   \____\___/|_| |_|_||_|  \__,_|\___\___|_|
#
#       ---  High-Performance Connectivity Tracer (HiPerConTracer)  ---
#                 https://www.nntb.no/~dreibh/hipercontracer/
# ==========================================================================
#
# High-Performance Connectivity Tracer (HiPerConTracer)
# Copyright (C) 2015-2026 by Thomas Dreibholz
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# Contact: dreibh@simula.no

import bz2
import datetime
import gzip
import io
import lzma
import zstandard

from enum   import Enum
from typing import Any, BinaryIO, Dict, Final, List, TextIO


class HopStatus(Enum):
   # ====== Status byte ===================================
   Unknown                   = 0

   # ====== ICMP responses (from routers) =================
   # NOTE: Status values from 1 to 199 have a given
   #       router IP in their HopIP result!

   # ------ TTL/Hop Count ---------------------------------
   TimeExceeded              = 1    # ICMP response

   # ------ Reported as "unreachable" ---------------------
   # NOTE: Status values from 100 to 199 denote unreachability
   UnreachableScope          = 100  # ICMP response
   UnreachableNetwork        = 101  # ICMP response
   UnreachableHost           = 102  # ICMP response
   UnreachableProtocol       = 103  # ICMP response
   UnreachablePort           = 104  # ICMP response
   UnreachableProhibited     = 105  # ICMP response
   UnreachableUnknown        = 110  # ICMP response

   # ====== No response  ==================================
   # NOTE: Status values from 200 to 254 have the destination
   #       IP in their HopIP field. However there is no response!
   Timeout                   = 200

   NotSentGenericError       = 210  # sendto() call failed (generic error)
   NotSentPermissionDenied   = 211  # sendto() error: EACCES
   NotSentNetworkUnreachable = 212  # sendto() error: ENETUNREACH
   NotSentHostUnreachable    = 213  # sendto() error: EHOSTUNREACH
   NotAvailableAddress       = 214  # sendto() error: EADDRNOTAVAIL
   NotValidMsgSize           = 215  # sendto() error: EMSGSIZE
   NotEnoughBufferSpace      = 216  # sendto() error: ENOBUFS

   # ====== Destination's response (from destination) =====
   # NOTE: Successful response destination is in HopIP field.
   Success                   = 255  # Success!

   # ------ Response received -----------------------------
   Flag_StarredRoute         = (1 << 8)   # Route with * (router did not respond)
   Flag_DestinationReached   = (1 << 9)   # Destination has responded


PROTOCOL_MAP : Final[Dict[str, str]] = {
   'i': 'ICMP',
   'u': 'UDP',
   't': 'TCP'
}


# ###### Is destination not reachable? ######################################
def statusIsUnreachable(hopStatus : HopStatus) -> bool:
   # Values 100 to 199 => the destination cannot be reached any more, since
   # a router on the way reported unreachability.
   return ( (hopStatus.value >= HopStatus.UnreachableScope.value) and
            (hopStatus.value < HopStatus.Timeout.value) )


# ###### Convert Unix time in nanoseconds to datetime #######################
def unixtime_to_datetime(unixTime : int) -> datetime.datetime:
   seconds : Final[int] = unixTime // 1_000_000_000
   dt      : Final[datetime.datetime] = \
      datetime.datetime.fromtimestamp(seconds, tz=datetime.UTC)
   return dt


# ###### Convert Unix time in nanoseconds to string #########################
def unixtime_to_string(unixTime : int) -> str:
   nanoseconds : Final[int] = unixTime % 1_000_000_000
   dt          : Final[datetime.datetime] = unixtime_to_datetime(unixTime)
   return dt.strftime(f'%Y-%m-%d %H:%M:%S.{nanoseconds:09d}')


# ###### Convert datetime to Unix time in nanoseconds #######################
def datetime_to_unixtime(timeStamp : datetime.datetime) -> int:
   seconds : Final[int] = int(timeStamp.timestamp())
   return seconds * 1_000_000_000 + timeStamp.microsecond * 1_000


# ###### Convert string to Unix time in nanoseconds #########################
def string_to_unixtime(timeString : str) -> int:
   # Split base date/time from sub-second fraction:
   parts : Final[List[str]] = timeString.strip().split('.')
   dt    : Final[datetime.datetime] = \
      datetime.datetime.fromisoformat(parts[0] + '+00:00').replace(tzinfo = datetime.UTC)
   seconds     : Final[int] = int(dt.timestamp())
   nanoseconds : int        = 0
   if len(parts) > 1:
      # Normalize subsecond string to 9 digits (nanoseconds):
      nanosecondsString : Final[str] = parts[1].ljust(9, '0')[:9]
      nanoseconds = int(nanosecondsString)
   return seconds * 1_000_000_000 + nanoseconds


# ###### Open HiPerConTracer results file ###################################
def openHiPerConTracerFile(hpctFileName : str) -> TextIO:
   # ====== Identify compression ============================================
   magic : bytes
   fh    : BinaryIO
   with open(hpctFileName, 'rb') as fh:
      magic = fh.read(6)

   # ====== Open file and provide uncompressed input stream =================
   if magic.startswith(b'\x1f\x8b'):
      return gzip.open(hpctFileName, 'rt', encoding='utf-8')
   elif magic.startswith(b'\xfd7zXZ'):
      return lzma.open(hpctFileName, 'rt', encoding='utf-8')
   elif magic.startswith(b'BZh'):
      return bz2.open(hpctFileName, 'rt', encoding='utf-8')
   elif magic.startswith(b'\x28\xb5\x2f\xfd'):
      fh = open(hpctFileName, 'rb')
      dc = zstandard.ZstdDecompressor()
      return io.TextIOWrapper(dc.stream_reader(fh), encoding='utf-8')
   else:
      return open(hpctFileName, 'rt', encoding='utf-8')
